### ISSUE TITLE
Setting the DocumentManager temporarily to read-only state

### ISSUE THREAD

**Issue by vmattila** [doctrine/mongodb-odm#1076]
I'd like to share an idea about something that could setting the DocumentManager to read-only state for certain operations. When read-only document manager is used, each read document would be automatically detached at the end. Still we could explicitly merge a document if they need to be updated later.

This would allow us to use same code to fetch sets of documents without knowing for what purpose the documents are actually used in other parts of the application.

The interface could be something similar to ORM's `::transactional()`, for example:

```
$dm->readOnly(function($dm) {
    $documents = SomeOtherHelperClass::findDocuments();
    // Do something read-only with the docs
});
```

Detaching would happen at the end of the callback function (as similar to ORM's `transactional()` does flushing) for all documents that were unknown to DocumentManager at the time it entered the read-only state.

**Comment by malarzm** [comment]
If I understand correctly then DocumentManager state would be snapshotted at the beginning and all documents added in read only mode would be detached later so DM's state will the same as in the beginning? It might be interesting to fetch a lot of objects with this and not care about penalty by using deferred implicit change policy if we don't want to change them anyway (manually detaching stuff would work as well but why not automate it) :)

**Comment by vmattila** [comment]
Yes, this is exactly what I was thinking. I'd say that detaching could be done as soon as the object is read so it would be also available for gc as soon as possible.

**Comment by malarzm** [comment]
While not on `DocumentManager` level, read-only mode will be possible in `Query` (and its `Builder`) after #1403 gets merged :)
